libgalois/include/katana/EntityTypeManager.h

```diff
@@ -0,0 +1,227 @@
+static constexpr EntityTypeID kInvalidEntityType =
+    std::numeric_limits<EntityTypeID>::max();
+/// A set of EntityTypeIDs
+using SetOfEntityTypeIDs =
+    std::bitset<std::numeric_limits<EntityTypeID>::max() + 1>;
+/// A map from EntityTypeID to a set of EntityTypeIDs
+using EntityTypeIDToSetOfEntityTypeIDsMap = std::vector<SetOfEntityTypeIDs>;
+/// A map from the atomic type name to its EntityTypeID
+/// (that does not intersect any other atomic type)
+using AtomicTypeNameToEntityTypeIDMap =
+    std::unordered_map<std::string, EntityTypeID>;
+/// A map from the atomic type's EntityTypeID to its name
+/// (that does not intersect any other atomic type)
+using EntityTypeIDToAtomicTypeNameMap =
+    std::unordered_map<EntityTypeID, std::string>;
+
+class EntityTypeManager {
+public:
+  EntityTypeManager() { Init(); }
+
+  EntityTypeManager(
+      EntityTypeIDToAtomicTypeNameMap&& atomic_entity_type_id_to_type_name,
+      EntityTypeIDToSetOfEntityTypeIDsMap&&
+          entity_type_id_to_atomic_entity_type_ids)
+      : atomic_entity_type_id_to_type_name_(
+            std::move(atomic_entity_type_id_to_type_name)),
+        entity_type_id_to_atomic_entity_type_ids_(
+            std::move(entity_type_id_to_atomic_entity_type_ids)) {
+    for (auto& type_id_name_pair : atomic_entity_type_id_to_type_name_) {
+      atomic_type_name_to_entity_type_id_.emplace(
+          std::make_pair(type_id_name_pair.second, type_id_name_pair.first));
+    }
+
+    size_t num_entity_types = entity_type_id_to_atomic_entity_type_ids_.size();
+    atomic_entity_type_id_to_entity_type_ids_.resize(num_entity_types);
+    for (size_t i = 0, ni = num_entity_types; i < ni; ++i) {
+      for (size_t j = 0, nj = num_entity_types; j < nj; ++j) {
+        if (entity_type_id_to_atomic_entity_type_ids_[i].test(j)) {
+          atomic_entity_type_id_to_entity_type_ids_[j].set(i);
+        }
+      }
+    }
+  }
+
+  EntityTypeManager(
+      EntityTypeIDToAtomicTypeNameMap&& atomic_entity_type_id_to_type_name,
+      AtomicTypeNameToEntityTypeIDMap&& atomic_type_name_to_entity_type_id,
+      EntityTypeIDToSetOfEntityTypeIDsMap&&
+          entity_type_id_to_atomic_entity_type_ids,
+      EntityTypeIDToSetOfEntityTypeIDsMap&&
+          atomic_entity_type_id_to_entity_type_ids)
+      : atomic_entity_type_id_to_type_name_(
+            std::move(atomic_entity_type_id_to_type_name)),
+        atomic_type_name_to_entity_type_id_(
+            std::move(atomic_type_name_to_entity_type_id)),
+        entity_type_id_to_atomic_entity_type_ids_(
+            std::move(entity_type_id_to_atomic_entity_type_ids)),
+        atomic_entity_type_id_to_entity_type_ids_(
+            std::move(atomic_entity_type_id_to_entity_type_ids)) {}
+
+  void Reset() {
+    atomic_entity_type_id_to_type_name_.clear();
+    atomic_type_name_to_entity_type_id_.clear();
+    entity_type_id_to_atomic_entity_type_ids_.clear();
+    atomic_entity_type_id_to_entity_type_ids_.clear();
+    Init();
+  }
```

Comment:
Is there a compelling reason to provide this as part of the public API? I would generally rather create a new object than reset and reuse. It's just clearer. (I understand why this is useful as a private method to implement assignment operators and destructors.

Reply:
Agree with Arthur.